cajita/unit_test/tstSparseDimPartitioner.hpp

```diff
@@ -0,0 +1,473 @@
+                    ranks_per_dim[d] )
+            {
+                int rand_num = std::rand() % size_tile_per_dim;
+                gt_partition_set[d].insert( rand_num );
+            }
+            gt_partition_set[d].insert( size_tile_per_dim );
+            int i = 0;
+            for ( auto it = gt_partition_set[d].begin();
+                  it != gt_partition_set[d].end(); ++it )
+            {
+                gt_partition[d][i++] = *it;
+            }
+        }
+    }
+
+    // broadcast the ground truth partition to all ranks
+    for ( int d = 0; d < 3; ++d )
+    {
+        MPI_Barrier( MPI_COMM_WORLD );
+        MPI_Bcast( gt_partition[d].data(), gt_partition[d].size(), MPI_INT, 0,
+                   MPI_COMM_WORLD );
+        MPI_Barrier( MPI_COMM_WORLD );
+    }
+
+    // init partitions (average partition)
+    std::array<std::vector<int>, 3> rec_partitions;
+    for ( int d = 0; d < 3; ++d )
+    {
+        int ele = size_tile_per_dim / ranks_per_dim[d];
+        int part = 0;
+        for ( int i = 0; i < ranks_per_dim[d]; ++i )
+        {
+            rec_partitions[d].push_back( part );
+            part += ele;
+        }
+        rec_partitions[d].push_back( size_tile_per_dim );
+    }
+
+    partitioner.initializeRecPartition( rec_partitions[0], rec_partitions[1],
+                                        rec_partitions[2] );
+
+    // basic settings for domain size and position
+    double cell_size = 0.1;
+    int pre_alloc_size = size_per_dim * size_per_dim;
+    std::array<double, 3> global_low_corner = { 1.2, 3.3, -2.8 };
+    std::array<double, 3> global_high_corner = {
+        global_low_corner[0] + cell_size * global_cells_per_dim[0],
+        global_low_corner[1] + cell_size * global_cells_per_dim[1],
+        global_low_corner[2] + cell_size * global_cells_per_dim[2] };
+
+    // use tile occupization info to compute the workload on MPI ranks
+    if ( use_tile2workload )
+    {
+        // randomly generate a fixed number of tiles on every MPI rank
+        auto tiles_view = generate_random_tiles(
+            gt_partition, cart_rank, size_tile_per_dim, occupy_num_per_rank );
+        // create a new sparseMap
+        auto global_mesh = createSparseGlobalMesh(
+            global_low_corner, global_high_corner, global_cells_per_dim );
+        auto sis =
+            createSparseMap<TEST_EXECSPACE>( global_mesh, pre_alloc_size );
+        // register selected tiles to the sparseMap
+        Kokkos::parallel_for(
+            "insert_tile_to_sparse_map",
+            Kokkos::RangePolicy<TEST_EXECSPACE>( 0, tiles_view.extent( 0 ) ),
+            KOKKOS_LAMBDA( int id ) {
+                sis.insertTile( tiles_view( id, 0 ), tiles_view( id, 1 ),
+                                tiles_view( id, 2 ) );
+            } );
+
+        // compute workload from a sparseMap and do partition optimization
+        partitioner.optimizePartition( sis, MPI_COMM_WORLD );
```

Comment:
Does initialize have to be called before optimize? If so and we always want to optimize, can we create a version of these functions that fuses them together?

Reply:
Only one time init before simulations start, and multiple times of optimization during the simulation. 
When optimizing, we will fix two dimensions and optimize the rest, so the partition information is needed in those two dimensions. That's why we do initialization before optimization, but only for once, as the results of the previous optimization can serve as the init state for the next round of optimization.